File: server/handlers/raregoods.py
from flask import render_template
from server.powers import get_system_power_info, is_system_anarchy
from server.tasks.raregoods import best_rare_goods
from server.tasks.tasks import task_description, is_task_own_strength, isPowersWeakness, isTaskACrime, systemNotes


def handle_rare_goods(request, database):
    """
    Handler for /results, when the task is rare goods

    Methods: GET

    Renders raregoods.html
    """
    system_name = request.args.get("system")
    system = request.args.get("system")
    task = request.args.get("taskName")
    power = request.args.get("power")

    powerInfo = get_system_power_info(system, database)
    controllingPower = powerInfo[1]
    systemState = powerInfo[0]
    rare_goods = best_rare_goods(system_name, database)
    return render_template(
        "tasks/raregoods.html",
        system=system,
        power=power,
        currentPower=controllingPower,
        currentState=systemState,
        isAnarchy="YES" if is_system_anarchy(system, database) else "NO",
        taskName=task,
        taskType="Trading",
        isIllegal="isn't",
        isOpposingWeakness=isPowersWeakness(power, task),
        taskDescription=task_description(task, power, system, powerInfo, database),
        systemNotes=systemNotes(power, system, database),
        raregoods=rare_goods,
        isOwnStrength="is" if is_task_own_strength(task, controllingPower) else "isn't"
    )

File: src/handlers/tasks/rareGoods.go
package tasks

import (
	"log"
	"maps"
	"math"
	"slices"

	"github.com/labstack/echo/v5"
	"niceygy.net/powerplay-assistant/src/database"
	"niceygy.net/powerplay-assistant/src/utils"
)

const rg_task_name string = "Deliver Rare Goods"
const rg_task_code string = "SRGD"
const rg_items_limit int = 15

type rare_good struct {
	Name     string
	System   string
	Station  string
	Distance float64
}

func HandleRareGoods(c *echo.Context) error {
	system := c.QueryParam("system")
	power := c.QueryParam("power")

	user_coords := database.GetSystemLocation(system)
	state, target_sys_owning_power := database.GetSystemPowerInfo(system)

	rows, err := database.Db.Query(`SELECT Raregoods.*, 
	` + database.CreateDistanceStatement(user_coords) + `
	FROM Raregoods
	INNER JOIN systems
		ON Raregoods.system_name = systems.system_name
	ORDER BY distance;`)

	if err != nil {
		log.Panic(err.Error())
	} else {
		defer rows.Close()
	}

	resultMap := map[float64]rare_good{}

	for rows.Next() {
		r := rare_good{}
		if err = rows.Scan(&r.Name, &r.System, &r.Station, &r.Distance); err != nil {
			log.Panic(err.Error())
		} else {
			r.Distance = math.Round(r.Distance)
		}
		diff_from_200 := math.Abs(r.Distance - 200)
		resultMap[diff_from_200] = r
		// result = append(result, r)
	}

	sorted_keys := slices.Sorted(maps.Keys(resultMap))
	var result []rare_good = make([]rare_good, 0)
	for i := 0; i < rg_items_limit; i++ {
		result = append(result, resultMap[sorted_keys[i]])
	}

	return c.HTML(200, utils.RenderTemplate("templates/tasks/raregoods.html", map[string]any{
		"system":             system,
		"power":              power,
		"currentPower":       target_sys_owning_power,
		"currentState":       state,
		"isAnarchy":          database.IsSystemAnarchy(system),
		"taskName":           rg_task_name,
		"taskType":           "Trading",
		"isIlligal":          "isn't",
		"isOpposingWeakness": utils.IsOpposingWeakness(rg_task_code, target_sys_owning_power),
		"taskDescription":    utils.TASK_DESCRIPTIONS[rg_task_code],
		"systemNotes":        database.GetExtraInfo(system),
		"raregoods":          result,
		"isOwnStrength":      utils.IsOwnStrength(rg_task_code, database.PowerFullToShort(power)),
	}))
}
